elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/QueryEngine.java

```diff
@@ -0,0 +1,126 @@
+ *                      | DataStoreTransaction |        +-------+  JpaQueryEngine  |
+ *                      |                      |        |       |                  |
+ *                      +----------+-----------+        |       +------------------+
+ *                                 |                    |
+ *                                 |                    |
+ *                          +------v------+             |       +----------------------+
+ *                          |             |             |       |                      |
+ *                          | QueryEngine +-------------+-------+ HibernateQueryEngine |
+ *                          |             |                     |                      |
+ *                          +------+------+                     +----------------------+
+ *                                 |
+ *                                 |
+ *             +-------------------------------------+
+ *             |                   |                 |
+ * +-----------v----------+   +----v----+   +--------v-------+
+ * |     Apache Drill     |   |   JPA   |   |    Hibernate   |
+ * +-----------+----------+   +----+----+   +--------+-------+
+ *             |                   |                 |
+ * +-----------v-------------------v-----------------v--------+
+ * |                         SQL Driver                       |
+ * +---------------+----------------------------+-------------+
+ *                 |                            |
+ *                 |                            |
+ *           +-----v-----+                  +---v----+
+ *           |           |                  |        |
+ *           |  NoSQL DB |                  | SQL DB |
+ *           |           |                  |        |
+ *           +-----------+                  +--------+
+ * </pre>
+ * Implementor must assume that {@link DataStoreTransaction} will never keep reference to any internal state of a
+ * {@link QueryEngine} object. This ensures the plugability of various {@link QueryEngine} implementations.
+ * <p>
+ * A concrete and complete {@link QueryEngine} must also have the corresponding implementations of
+ * <ol>
+ *     <li> a {@link Query} that can be understood by a particular persistent storage or storage client
+ *     <li> a {@link WhereFilterExpression} that maps an Elide WHERE filter clause to an equivalent query constraint,
+ *          which can be interpreted by the particular persistent storage
+ *     <li> a {@link HavingFilterExpression} that maps an Elide WHERE filter clause to an equivalent query constraint,
+ *          which can be interpreted by the particular persistent storage
+ *     <li> one or more {@link FilterExpressionVisitor}s that can split an arbitrary {@link FilterExpression} into
+ *          aforementioned {@link WhereFilterExpression} and {@link HavingFilterExpression}
+ * </ol>
+ */
+public interface QueryEngine {
```

Comment:
A couple thoughts here.  I would make QueryEngine just an interface with a single method:
``` Iterable<Object> executeQuery(Query query);```.   However, I would make Query an actual object with fields and a lombok builder or something along those lines to make it easier to construct.